src/Weeder.hs

```diff
@@ -139,6 +140,10 @@ instance Show Declaration where
     declarationStableName
 
 
+-- TODO maybe we can make this faster by only hashing the location.
+instance Hashable Declaration where
+  hashIO32 d s = hashIO32 (declarationStableName d) s
```

Comment:
This is definitely not fast, and could be why this approach doesn't work.

Reply:
Could be worth trying the [`Uniquable`](https://hackage.haskell.org/package/ghc-9.6.6/docs/GHC-Types-Unique.html#t:Uniquable) instance of `Module` and `OccName` to produce hashes.

```
Note [The Unique of an OccName]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
They are efficient, because FastStrings have unique Int# keys.  We assume
this key is less than 2^24, and indeed FastStrings are allocated keys
sequentially starting at 0.

So we can make a Unique using
        mkUnique ns key  :: Unique
where 'ns' is a Char representing the name space.  This in turn makes it
easy to build an OccEnv.
-}
```

(or maybe even better: the `Uniquable` instance of the original `Name`s we derive the declarations from with `nameToDeclaration`)